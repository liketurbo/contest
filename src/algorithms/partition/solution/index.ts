import LinkedListNode from "../../../dataTypes/LinkedList/solution/LinkedListNode"

const partition = (node: LinkedListNode<number>, part: number) => {
  let startNode: LinkedListNode<number> | null = node
  let endNode: LinkedListNode<number> | null = node

  while (node) {
    const curNode = node

    node = node.next as LinkedListNode<number>

    if (curNode.val < part) {
      curNode.next = startNode
      startNode = curNode
    } else {
      endNode.next = curNode
      endNode = curNode
    }
  }

  endNode.next = null

  return startNode
}

export default partition
