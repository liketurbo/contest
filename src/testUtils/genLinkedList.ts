import LinkedListNode from "../dataTypes/LinkedList/solution/LinkedListNode"

const genLinkedList = <T>(arr: T[]) => {
  const dummyNode = new LinkedListNode(null) as LinkedListNode<unknown>

  let curNode = dummyNode

  arr.forEach((val) => {
    const newNode = new LinkedListNode(val)

    curNode.next = newNode
    curNode = newNode
  })

  return dummyNode.next
}

export default genLinkedList
