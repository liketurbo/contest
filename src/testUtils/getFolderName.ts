import path from "path"

const getCurrentFolder = () => {
  const err = new Error()

  const re = /(?<=\().+(?=:\d+:)/u

  const currentFile = (re.exec(err.stack?.split("\n")[3] as string) || [])[0]

  const currentFolder = path.dirname(currentFile)

  return currentFolder
}

const getFolderName = (lvl = 2) => {
  if (lvl < 1) throw new Error("Level should be at least one.")

  const lvls = lvl === 1 ? "." : "../".repeat(lvl - 1)

  const currentFolder = getCurrentFolder()

  return path.basename(path.join(currentFolder, lvls))
}

export default getFolderName
