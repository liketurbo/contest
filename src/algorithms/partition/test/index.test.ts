import LinkedListNode from "../../../dataTypes/LinkedList/solution/LinkedListNode"
import getFolderName from "../../../testUtils/getFolderName"
import genLinkedList from "../../../testUtils/getLinkedList"
import partition from "../solution"

describe(getFolderName(), () => {
  test("book example", () => {
    const list = genLinkedList([3, 5, 8, 5, 10, 2, 1]) as LinkedListNode<number>

    expect(partition(list, 5)).toMatchSnapshot()
  })
})
