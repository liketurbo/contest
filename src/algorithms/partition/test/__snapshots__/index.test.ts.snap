// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`partition book example 1`] = `
LinkedListNode {
  "next": LinkedListNode {
    "next": LinkedListNode {
      "next": LinkedListNode {
        "next": LinkedListNode {
          "next": LinkedListNode {
            "next": LinkedListNode {
              "next": null,
              "val": 10,
            },
            "val": 5,
          },
          "val": 8,
        },
        "val": 5,
      },
      "val": 3,
    },
    "val": 2,
  },
  "val": 1,
}
`;
